SICP_book/chapter_1_2.metta

```diff
@@ -0,0 +1,645 @@
+; Extension is used for several examples which requires random and current-inexact-milliseconds functions from Scheme
+!(extend-py! additional_funcs)
+
+(= (inc $x) (+ $x 1))
+(= (dec $x) (- $x 1))
+
+(= (sqr $x) (* $x $x))
+
+(= (>= $x $y) (or (> $x $y) (== $x $y)))ex
+(= (<= $x $y) (or (< $x $y) (== $x $y)))
+
+; Since +,/,*,- in metta takes only two arguments as input:
+(= (+3 $a $b $c) (+ $a (+ $b $c)))
+(= (*3 $a $b $c) (* $a (* $b $c)))
+
+; Recursive factorial evaluation
+(= (factorial $n)
+    (if (== $n 1)
+    1
+    (* $n (factorial (dec $n)))))
+
+!(assertEqual
+    (factorial 5)
+    120)
+
+; Iterative factorial evaluation
+(= (ifactorial $n)
+    (fact-iter 1 1 $n))
+
+(= (fact-iter $product $counter $max-count)
+    (if (> $counter $max-count)
+        $product
+        (fact-iter (* $counter $product) (+ $counter 1) $max-count)))
+
+!(assertEqual
+    (ifactorial 5)
+    120)
+
+; Currently there is no cond in metta, so we are just using several ifs.
```

Comment:
There is a `case` which is rather similar and can be used in this case. But not sure it it important to use one construction or another. You can use it in a following manner to replace `cond`:
```
(case True (
  ((== $n 0) 0)
  ((== $n 1) 1)
  ($_ (+ (fib (dec $n)) (fib (- $n 2)))) ))
```

Reply:
I am sorry, I have found that this `case` example doesn't work because `case` doesn't evaluates patterns. I don't see anything wrong with using `if` here though.